pkg/http/listener.go

```diff
@@ -0,0 +1,284 @@
+ * limitations under the License.
+ */
+
+package http
+
+import (
+	"crypto/tls"
+	"errors"
+	"net"
+	"strings"
+	"sync"
+	"time"
+)
+
+// HTTP methods.
+var methods = []string{
+	"OPTIONS",
+	"GET",
+	"HEAD",
+	"POST",
+	"PUT",
+	"DELETE",
+	"TRACE",
+	"CONNECT",
+	"PRI", // HTTP 2 method
+}
+
+// maximum length of above methods + one space.
+var methodMaxLen = getMethodMaxLen() + 1
+
+func getMethodMaxLen() int {
+	maxLen := 0
+	for _, method := range methods {
+		if len(method) > maxLen {
+			maxLen = len(method)
+		}
+	}
+
+	return maxLen
+}
+
+func isHTTPMethod(s string) bool {
+	for _, method := range methods {
+		if s == method {
+			return true
+		}
+	}
+
+	return false
+}
+
+type acceptResult struct {
+	conn net.Conn
+	err  error
+}
+
+// httpListener - HTTP listener capable of handling multiple server addresses.
+type httpListener struct {
+	mutex                  sync.Mutex         // to guard Close() method.
+	tcpListeners           []*net.TCPListener // underlaying TCP listeners.
+	acceptCh               chan acceptResult  // channel where all TCP listeners write accepted connection.
+	doneChs                []chan struct{}    // done channels for each goroutine for TCP listeners.
+	tlsConfig              *tls.Config        // TLS configuration
+	tcpKeepAliveTimeout    time.Duration
+	readTimeout            time.Duration
+	writeTimeout           time.Duration
+	updateBytesReadFunc    func(int)                           // function to be called to update bytes read in bufConn.
+	updateBytesWrittenFunc func(int)                           // function to be called to update bytes written in bufConn.
+	errorLogFunc           func(error, string, ...interface{}) // function to be called on errors.
+}
+
+// start - starts separate goroutine for each TCP listener.  A valid insecure/TLS HTTP new connection is passed to httpListener.acceptCh.
+func (listener *httpListener) start() {
+	listener.acceptCh = make(chan acceptResult)
+
+	// Closure to accept single connection.
+	acceptTCP := func(tcpListener *net.TCPListener) (net.Conn, error) {
+		tcpConn, err := tcpListener.AcceptTCP()
+		if err != nil {
+			return nil, err
+		}
+
+		// Tune accepted TCP connection.
+		tcpConn.SetKeepAlive(true)
+		tcpConn.SetKeepAlivePeriod(listener.tcpKeepAliveTimeout)
+
+		// Wrap TCPConn using bufConn to peek HTTP method.
+		bufconn := newBufConn(tcpConn, listener.readTimeout, listener.writeTimeout,
+			listener.updateBytesReadFunc, listener.updateBytesWrittenFunc)
+
+		// Peek bytes of maximum length of all HTTP methods.
+		var data []byte
+		if data, err = bufconn.Peek(methodMaxLen); err != nil {
+			if listener.errorLogFunc != nil {
+				listener.errorLogFunc(err,
+					"Error in reading from new connection %s at server %s",
+					bufconn.RemoteAddr(), bufconn.LocalAddr())
+			}
+			bufconn.Close()
+			return nil, nil
+		}
+
+		// Return bufconn if read data is a valid HTTP method.
+		tokens := strings.SplitN(string(data), " ", 2)
```

Comment:
Have you profiled this code? @balamurugana - i wonder if we can avoid using `SplitN()` or simply use bytes directly instead of strings for this. We need to avoid allocation.